### PR TITLE
implemented suggested changes, borrow-repay feature for erc20

### PR DESCRIPTION
**Pull request by chinmay0402**
- all ether/erc20 tokens sent to the contract now get sent to the user
- implemented borrow and repay for erc20 tokens using ether as collateral
- removed _redeemType parameter for withdraw functions

### CODE REVIEW

File: contracts/CompoundMiddleContract.sol [hunk 1]
```diff
@@ -71,7 +71,11 @@ interface PriceFeed {
 }
```
Comment (pradyuman-verma) [inline contracts/CompoundMiddleContract.sol]: Let's move all the interfaces into a separate file and import them here.

File: contracts/CompoundMiddleContract.sol [hunk 2]
```diff
@@ -103,48 +107,31 @@ contract CompoundMiddleContract {
...
+        uint256 amountOfEthreceived = _redeemAmount *
+            cEth.exchangeRateCurrent();
+
+        console.log(
+            "Redeemed %s cEth for %s ether",
+            _redeemAmount,
+            amountOfEthreceived
+        );
 
         require(redeemResult == 0, "ERROR WHILE REDEEMING");
```
Comment (pradyuman-verma) [inline contracts/CompoundMiddleContract.sol]: This statement should be just below where we are calculating `reedemResult`.
```suggestion
        require(redeemResult == 0, "ERROR WHILE REDEEMING");
```

File: contracts/CompoundMiddleContract.sol [hunk 3]
```diff
@@ -200,11 +189,10 @@ contract CompoundMiddleContract {
         console.log("Liquidity available", liquidity); // liquidity is the USD value borrowable by the user, before it reaches liquidation
 
         // borrow
-        uint256 amountToBorrowInWei = 2000000000000000; // CHANGE
-        cEth.borrow(amountToBorrowInWei);
+        cEth.borrow(_amountToBorrowInWei);
```
Comment (pradyuman-verma) [inline contracts/CompoundMiddleContract.sol]:
```suggestion
        require(cEth.borrow(_amountToBorrowInWei) == 0, "Borrow failed!");
```

File: contracts/CompoundMiddleContract.sol [hunk 4]
```diff
@@ -213,30 +201,31 @@ contract CompoundMiddleContract {
...
-    function paybackEth(address _cEtherAddress, uint256 amount)
-        public
-        returns (bool)
-    {
+    function paybackEth(
+        address _cEtherAddress,
+        uint256 amount,
+        uint256 gas
+    ) public returns (uint256) {
         CEth cEth = CEth(_cEtherAddress);
```
Comment (pradyuman-verma) [inline contracts/CompoundMiddleContract.sol]: This function should be payable.

File: contracts/CompoundMiddleContract.sol [hunk 5]
```diff
@@ -213,30 +201,31 @@ contract CompoundMiddleContract {
...
-        cEth.repayBorrow{value: amount}();
+        cEth.repayBorrow{value: amount, gas: gas}();
```
Comment (pradyuman-verma) [inline contracts/CompoundMiddleContract.sol]:
```suggestion
        cEth.repayBorrow{value: msg.value, gas: gas}();
```

File: contracts/CompoundMiddleContract.sol [hunk 6]
```diff
@@ -247,65 +182,116 @@ contract CompoundMiddleContract {
         underlying.approve(_cErc20Contract, _numTokensToSupply);
```
Comment (pradyuman-verma) [inline contracts/CompoundMiddleContract.sol]: I think we have to use `transferFrom` function as well to transfer the user's ERC20 tokens to address(this).

Also, let's use the safeERC20 library.

Comment (chinmay0402) [reply contracts/CompoundMiddleContract.sol]: Oh right, need to transfer to the contract first.

File: contracts/CompoundMiddleContract.sol [hunk 7]
```diff
@@ -191,121 +119,177 @@ contract CompoundMiddleContract {
...
         // create references to the contracts on mainnet
-        Erc20 underlying = Erc20(_erc20Contract);
+        Erc20 token = Erc20(_erc20Contract);
         CErc20 cToken = CErc20(_cErc20Contract);
 
+        token.transferFrom(msg.sender, address(this), _numTokensToSupply);
+
         // Approve transfer on the ERC20 contract
-        underlying.approve(_cErc20Contract, _numTokensToSupply);
+        token.approve(_cErc20Contract, _numTokensToSupply);
```
Comment (pradyuman-verma) [inline contracts/CompoundMiddleContract.sol]: ~`Approve` first then `transferFrom`~

Comment (chinmay0402) [reply contracts/CompoundMiddleContract.sol]: As in `approve` the contract to spend the user's tokens? Can we do that from inside this contract?

Comment (pradyuman-verma) [reply contracts/CompoundMiddleContract.sol]: It would be better to do this step using script.

Comment (pradyuman-verma) [reply contracts/CompoundMiddleContract.sol]: yes, we are approving the `_cErc20Contract ` to use funds

File: contracts/CompoundMiddleContract.sol [hunk 8]
```diff
@@ -191,121 +119,177 @@ contract CompoundMiddleContract {
...
 
         // supply the tokens to Compound and mint cTokens
-        uint mintResult = cToken.mint(_numTokensToSupply);
+        uint256 mintResult = cToken.mint(_numTokensToSupply);
```
Comment (pradyuman-verma) [inline contracts/CompoundMiddleContract.sol]: How can you be sure that the ERC20 tokens got minted?

File: contracts/CompoundMiddleContract.sol [hunk 9]
```diff
@@ -187,125 +119,182 @@ contract CompoundMiddleContract {
...
+
+        // transfer user's tokens to contract
+        token.safeTransferFrom(msg.sender, address(this), _repayAmount);
+
+        // approve Compound to spend erc20 tokens
+        token.safeApprove(_cErc20Address, _repayAmount);
+
+        // repay borrow
+        cToken.repayBorrow(_repayAmount);
+
```
Comment (pradyuman-verma) [inline contracts/CompoundMiddleContract.sol:307]: Are you sure in here, that repayBorrow succeeds in all cases?

File: contracts/CompoundMiddleContract.sol (hunk 8 above)
Comment (chinmay0402) [reply contracts/CompoundMiddleContract.sol]: Handled

File: contracts/CompoundMiddleContract.sol (hunk 9 above)
Comment (chinmay0402) [reply contracts/CompoundMiddleContract.sol:307]: handled

File: contracts/CompoundMiddleContract.sol [hunk 10]
```diff
@@ -185,127 +111,206 @@ contract CompoundMiddleContract {
...
         // create references to the contracts on mainnet
-        Erc20 underlying = Erc20(_erc20Contract);
+        IERC20 token = IERC20(_erc20Contract);
         CErc20 cToken = CErc20(_cErc20Contract);
 
+        token.safeTransferFrom(msg.sender, address(this), _numTokensToSupply);
         // Approve transfer on the ERC20 contract
-        underlying.approve(_cErc20Contract, _numTokensToSupply);
+       
+        token.safeIncreaseAllowance(_cErc20Contract, _numTokensToSupply);
```
Comment (pradyuman-verma) [inline contracts/CompoundMiddleContract.sol]:
```suggestion
        token.safeApprove(_cErc20Contract, _numTokensToSupply);
```

**Pull request merged**